crates/y-sweet-core/src/sync_kv.rs

```diff
@@ -50,14 +50,14 @@ impl SyncKv {
     }
 
     fn mark_dirty(&self) {
-        if !self.dirty.load(Ordering::Relaxed) && !self.shutdown.load(Ordering::SeqCst) {
-            self.dirty.store(true, Ordering::Relaxed);
+        if !self.is_dirty() && !self.shutdown.load(Ordering::SeqCst) {
+            self.dirty.store(true, Ordering::SeqCst);
             (self.dirty_callback)();
         }
```

Comment:
Good catch but technically there is still a bug here, because the value of `self.dirty` could change between the read and the write. We could use an atomic swap instead, like this:

```rust
if !self.shutdown.load(Ordering::SeqCst) {
    if !self.dirty.swap(true, Ordering::SeqCst) {
        (self.dirty_callback)();
    }
}
```